### ISSUE TITLE
Modify the 'cf_role' attribute check for recent IOOS Profile 1.2 changes

### ISSUE THREAD

**Issue by mwengren** [ioos/compliance-checker#828]
@daltonkell We recently made some changes to IOOS Profile 1.2 to relax the requirement that the `cf_role` attribute be attached to the 'Platform Variable'.  See the first two attributes in the attribute table: https://ioos.github.io/ioos-metadata/ioos-metadata-profile-v1-2.html#platform for new rules.  

A good example dataset where this is the case is PacIOOS' AWS-HIMB dataset: https://pae-paha.pacioos.hawaii.edu/erddap/tabledap/AWS-HIMB.das.  

It includes both a `station_name` variable that includes the `cf_role` attribute and a `platform1` variable referenced by the `geophysical_variable:platform` attributes. 

Currently CC 4.3.3 gives this warning message:

Name | Reasoning
--|--
cf_role | Platform variable "platform1" must have a  cf_role attribute with one of the values ['profile_id', 'timeseries_id',  'trajectory_id']

Can we relax or remove this test for an upcoming patch release?

**Comment by daltonkell** [comment]
@mwengren 

So that dataset has variables:

```
...
  platform1 {
    Int32 _FillValue -999;
    String long_name "AWS-HIMB: HIMB Weather Station: Moku o Loe (Coconut Island), Oahu, Hawaii";
    String ncei_code "32FS";
    String short_name "AWS-HIMB";
  }
...
  station_name {
    Int32 _FillValue -999;
    String cf_role "timeseries_id";
    String ioos_category "Identifier";
    String long_name "AWS-HIMB: HIMB Weather Station: Moku o Loe (Coconut Island), Oahu, Hawaii";
    String short_name "AWS-HIMB";
  }
...
```

and the new changes in the Profile are

```
cf_role may be applied to the ‘Platform Variable’, as indicated by geophysical_variable:platform, but it may
also be an independent variable. To comply with the single platform per dataset rule of the IOOS
metadata Profile, the cf_role variable will typically have a dimension of 1, unless it is a TimeSeries dataset
following the ‘TimeSeries - multiple-station’ format.
```

In conclusion, `cf_role` must be present on _a variable_, but that variable need not be a `platform` variable. The variable to which `cf_role` is assigned must have a dimension of `1`, unless the dataset is following the __TimeSeries - multiple-station__ format. The above dataset would check out as all good in this case. Am I understanding that correctly?

**Comment by mwengren** [comment]
Looks good to me!

**Comment by daltonkell** [comment]
@mwengren revisiting this now -- I may be doing this to myself, but I'm a bit confused by some wording in the specification.

Say a dataset has `featureType=TimeSeriesProfile - single station`. Thus there must be two variables to describe each `cf_role` - one for `timeseries_id` and one for `profile_id`, as is consistent with CF. In this case, it is possible (likely?) that the variable with `cf_role=profile_id` will have a dimension have greater than `1`, as some stations may take many different profiles. If my logic is correct, such a dataset would fail the mandate where the `cf_role` variable has a dimension of `1`. Isn't this dataset valid in reality though? Is having multiple profiles in one file not advised?

**Comment by mwengren** [comment]
Yes, we discussed the different possibilities for the `cf_role` attribute dimensions earlier.  Took me a minute, but I dug it up in: https://github.com/ioos/compliance-checker/issues/748#issuecomment-606659685.  

I oversimplified in my comment above, for `featureType=timeSeriesProfile|trajectoryProfile` the `cf_role=profile_id` variable would both be allowed to be > 1 to allow multiple profiles per station or glider/ship, respectively.   HTH.

**Comment by daltonkell** [comment]
@mwengren With the updates I've written, the above dataset no longer shows that error message. Are there any other datasets you recommend testing out?

**Comment by mwengren** [comment]
I think this NANOOS one also is in the same situation: http://data.nanoos.org/erddap/tabledap/OOI_CE09OSSM.html

**Comment by daltonkell** [comment]
@mwengren testing on the NANOOS dataset, I get two messages of interest:

```
CF Discrete Sampling Geometry Compliance
* Dimension length of the variable with cf_role='timeseries_id (the 'station' dimension) is 8. Note that the IOOS profile restricts
 timeSeries datasets with multiple features to share the same lat/lon position (ie. to exist on the same platform). Datasets that i
nclude multiple platforms are not valid and will cause harvesting errors.

...

cf_role variables           # ----------> this is the result of the updated code
* Dimension length of non-platform variable with cf_role=timeseries_id  (the 'station' dimension) is 8. The IOOS Profile restricts 
timeSeries and timeSeries - single station datasets with multiple features to share the same lat/lon position (i.e. to exist on the
 same platform). Datasets that include multiple platforms are not valid and will cause harvesting errors.
```

This is what I expected, since the dataset has `featureType: timeseries` and not `featureType: timeseries - multiple station`. However, the CF DSG check will probably also need to be altered to incorporate these new relaxed rules (it currently doesn't look for `timeseries - multiple station` or `timeseries - single station`, the new caveats that are now in the spec). Are you on board with that?

Actually, now that I think of it, the new code should simply replace the current `CF DSG` check. It's checking the same things, only more thoroughly now that we have these added conditions.

**Comment by mwengren** [comment]
Those `featureType=timeSeries` variations (`multiple station|single station`) won't actually appear in the datasets, they're meant to be more conceptual for documentation purposes.  Below the table in the [Profile](https://ioos.github.io/ioos-metadata/ioos-metadata-profile-v1-2.html#platform), it's explained a little more:

> TimeSeries - multiple station: A DSG TimeSeries featureType dataset with a ‘station dimension’ of greater than 1

So we should stick with the original rules and messages we came up with (the ones based on `featureType` value).  To be clear, the only valid values for that attribute will be those in the table in [Chap 9.1](http://cfconventions.org/cf-conventions/cf-conventions.html#discrete-sampling-geometries).  

Sorry, I wasn't quite clear in responding to your earlier https://github.com/ioos/compliance-checker/issues/828#issuecomment-659562389, in the case of what we call `TimeSeriesProfile - multiple station`, we're just classifying the dataset that way because it's the one exception to the case where the `cf_role=station_id` variable can have dimension > 1.  I think the rules here https://github.com/ioos/compliance-checker/issues/748#issuecomment-606659685 still apply.

On the test results, it doesn't make sense to have repeated messages/results, agreed.  I think the first message and test name above looks better to me, without knowing implications on the code.

**Comment by daltonkell** [comment]
@mwengren so just be be crystal:

 - If a user selects `featureType=timeseries` and gives a `cf_role=timeseries_id` variable with dimension > 1, it's a failed check
 - If a user selects `featureType=timeseriesProfile` and gives a `cf_role=timseries_id` variable with dimension > 1, it's a failed check (profile dimension can be > 1, but not less than 1)
 - If a user selects `featureType=trajectory` and gives a `cf_role=trajectory_id` variable with dimension > 1, it's a failed check
 - If a user selects `featureType=trajectoryProfile` and gives `cf_role=trajectory_id` variable with dimension > 1, it's a failed check (profile dimension can be > 1, but not less than 1)
 - If a user selects `featureType=profile` and gives `cf_role=profile_id` variable with dimension > 1, it's a failed check
 - If a user selects `featureType=point`, default pass

**Comment by mwengren** [comment]
That looks right to me, depending on what's meant by failure in bullet 1.

For the first bullet case, we allow the `timeseries_id` dimension to be > 1 for NDBC/GTS harvesting purposes assuming the stations all share the same lat/long, so the message that's reported from the check should be similar to the one you quoted above, and it should pass those NDBC/GTS checks for the DSG dimension part of the check, anyway.  Example message:

```
Dimension length of the variable with cf_role='timeseries_id (the 'station' dimension) is 8. Note that the IOOS profile restricts
 timeSeries datasets with multiple features to share the same lat/lon position (ie. to exist on the same platform). Datasets that include multiple platforms are not valid and will cause harvesting errors.
```
Does that make sense?

**Comment by daltonkell** [comment]
Got it, so the test is going to "fail" always (only way to show the message), but it still may be considered a valid dimension size by the specification.
